File: src/main/leetcode/p208/trie/Trie.java
package p208.trie;

class TrieNode {
    private TrieNode[] children;
    private boolean end = false;
    // Initialize your data structure here.
    public TrieNode() {
        this.children = new TrieNode[26];
    }

    public void insert(char[] words, int i) {
        if (i >= words.length - 1) {
            this.end = true;
            return;
        }

        int j = i + 1;
        char nxt = words[j];
        TrieNode child = children[nxt - 'a'];
        if (child == null) {
            child = new TrieNode();
            children[nxt - 'a'] = child;
        }
        child.insert(words, j);
    }

    public boolean startsWith(char[] words, int i) {
        if(i >= words.length - 1) {
            return true;
        }
        int j = i + 1;
        char nxt = words[j];
        TrieNode child = children[nxt - 'a'];
        if(child == null) {
            return false;
        }
        return child.startsWith(words, j);
    }

    public boolean search(char[] words, int i) {
        if(i >= words.length - 1) {
            return this.end;
        }
        int j = i + 1;
        char nxt = words[j];
        TrieNode child = children[nxt - 'a'];
        if(child == null) {
            return false;
        }
        return child.search(words, j);
    }
}

public class Trie {
    private TrieNode root;

    public Trie() {
        root = new TrieNode();
    }

    // Inserts a word into the trie.
    public void insert(String word) {
        root.insert(word.toCharArray(), -1);
    }

    // Returns if the word is in the trie.
    public boolean search(String word) {
        return root.search(word.toCharArray(), -1);
    }

    // Returns if there is any word in the trie
    // that starts with the given prefix.
    public boolean startsWith(String prefix) {
        return root.startsWith(prefix.toCharArray(), -1);
    }

    public static void main(String[] args) {
        Trie trie = new Trie();
        trie.insert("ab");
        System.out.println(trie.search("a"));
        System.out.println(trie.startsWith("a"));
    }
}

// Your Trie object will be instantiated and called as such:
// Trie trie = new Trie();
// trie.insert("somestring");
// trie.search("key");

File: src/main/leetcode/set0000/set200/set200/p208/Trie.scala
package set0000.set200.set200.p208

class Trie() {

  /** Initialize your data structure here. */

  class Node {
    val children = Array.fill[Node](26)(null)
    var leaf = false

    def insert(word: String): Unit = {
      if (word.isEmpty) {
        leaf = true
      } else {
        val idx = word.head - 'a'
        if (children(idx) == null) {
          children(idx) = new Node
        }
        val child = children(idx)
        child.insert(word.tail)
      }
    }

    def search(word: String): Boolean = {
      if (word.isEmpty) {
        leaf
      } else {
        val idx = word.head - 'a'
        children(idx) != null && children(idx).search(word.tail)
      }
    }

    def startWith(prefix: String): Boolean = {
      if (prefix.isEmpty) {
        true
      } else {
        val idx = prefix.head - 'a'
        children(idx) != null && children(idx).startWith(prefix.tail)
      }
    }
  }

  val root = new Node

  /** Inserts a word into the trie. */
  def insert(word: String) {
    root.insert(word)
  }

  /** Returns if the word is in the trie. */
  def search(word: String): Boolean = {
    root.search(word)
  }

  /** Returns if there is any word in the trie that starts with the given prefix. */
  def startsWith(prefix: String): Boolean = {
    root.startWith(prefix)
  }

}
